File: app/src/main/java/com/przemolab/oknotifier/services/ContestIntentService.java
package com.przemolab.oknotifier.services;

import android.app.IntentService;
import android.appwidget.AppWidgetManager;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.support.annotation.Nullable;

import com.przemolab.oknotifier.Constants;
import com.przemolab.oknotifier.models.Contestant;
import com.przemolab.oknotifier.widgets.ContestWidgetDataProvider;
import com.przemolab.oknotifier.widgets.ContestWidgetProvider;

public class ContestIntentService extends IntentService {

    public static String ACTION_UPDATE_CONTESTS_WIDGETS = "update_contests_widgets";

    public ContestIntentService() {
        super(ContestIntentService.class.getName());
    }

    @Override
    protected void onHandleIntent(@Nullable Intent intent) {
        if (intent != null) {
            final String action = intent.getAction();
            if (ACTION_UPDATE_CONTESTS_WIDGETS.equals(action)) {
                handleActionUpdateRecipesWidgets();
            }
        }
    }

    public static void startActionUpdateContestWidgets(Context context) {
        Intent intent = new Intent(context, ContestIntentService.class);
        intent.setAction(ACTION_UPDATE_CONTESTS_WIDGETS);
        context.startService(intent);
    }

    private void handleActionUpdateRecipesWidgets() {
        SharedPreferences sharedPreferences = getSharedPreferences(Constants.SharedPreferences.Name, Context.MODE_PRIVATE);
        ContestWidgetDataProvider dataProvider = new ContestWidgetDataProvider(sharedPreferences);
        Contestant bestContestant = dataProvider.getBestContestant();

        AppWidgetManager appWidgetManager = AppWidgetManager.getInstance(this);
        int[] appWidgetIds = appWidgetManager.getAppWidgetIds(new ComponentName(this, ContestWidgetProvider.class));

        // update all widgets
        ContestWidgetProvider.updateAppWidgets(this, appWidgetManager, appWidgetIds, bestContestant);
    }
}

File: app/src/main/java/com/przemolab/oknotifier/services/ContestIntentService.kt
package com.przemolab.oknotifier.services

import android.app.IntentService
import android.appwidget.AppWidgetManager
import android.content.ComponentName
import android.content.Context
import android.content.Intent

import com.przemolab.oknotifier.Constants
import com.przemolab.oknotifier.widgets.ContestWidgetDataProvider
import com.przemolab.oknotifier.widgets.ContestWidgetProvider

class ContestIntentService : IntentService(ContestIntentService::class.java.name) {

    override fun onHandleIntent(intent: Intent?) {
        if (intent != null) {
            val action = intent.action
            if (ACTION_UPDATE_CONTESTS_WIDGETS == action) {
                handleActionUpdateRecipesWidgets()
            }
        }
    }

    private fun handleActionUpdateRecipesWidgets() {
        val sharedPreferences = getSharedPreferences(Constants.SharedPreferences.Name, Context.MODE_PRIVATE)
        val dataProvider = ContestWidgetDataProvider(sharedPreferences)
        val bestContestant = dataProvider.bestContestant

        val appWidgetManager = AppWidgetManager.getInstance(this)
        val appWidgetIds = appWidgetManager.getAppWidgetIds(ComponentName(this, ContestWidgetProvider::class.java))

        // update all widgets
        ContestWidgetProvider.updateAppWidgets(this, appWidgetManager, appWidgetIds, bestContestant)
    }

    companion object {

        var ACTION_UPDATE_CONTESTS_WIDGETS = "update_contests_widgets"

        fun startActionUpdateContestWidgets(context: Context) {
            val intent = Intent(context, ContestIntentService::class.java)
            intent.action = ACTION_UPDATE_CONTESTS_WIDGETS
            context.startService(intent)
        }
    }
}
